docs/easyrtc_configuration.md
EasyRTC: Configuration
======================

---
**OLD FILE - See easyrtc_server_configuration.md**

---

config.js
---------

Configuration of EasyRTC is done within config.js which is located in the EasyRTC root folder. It should be edited within your favorite text editor. Any changes will take effect the next time the server is started.


Ports
-----
- **httpPort** - The port which the http server will listen on. Commonly this is 80 or 8080.

- Note 1: Any ports which are opened may require firewall and NAT rules to be adjusted.

- Note 2: With Linux, choosing a port below 1024 may require the server be started with root privileges.

- Note 3: If sslEnable is true and sslForwardFromHttp if false, the http server will not be started, thus the port will not be used.


SSL
---

- **sslEnable** - Enables SSL for both https and web sockets. [true|false]
- **sslPort** - The port which the https server will listen on. Commonly this is 443 or 8443.
- **sslForwardFromHttp** - Forces non-secure connections to be forwarded to the secure site. [true|false]
- **sslKeyFile** - Local file location of private SSL key
- **sslCertFile** - Local file location of public SSL certificate

- Note 1: When SSL is enabled, all server traffic must use it. Optionally, the http port can be used to forward requests to use SSL.

- Todo: Document process of creating self-signed key


Experimental Stun Server
------------------------

- **experimentalStunServerEnable** - Enables the integrated STUN server. [true|false]
- **experimentalStunServerAddr0** - IP or domain pointing to server.
- **experimentalStunServerAddr1** - Second IP or domain pointing to server.
- **experimentalStunServerPort0** - First port number to listen on. Defaults to "3478".
- **experimentalStunServerPort1** - Second port number to listen on. Defaults to "3479".

- Note 1: The integrated stun server is useful in cases where there is no outside internet connection, or corporate policies / firewalls restrict outside connections.

- Note 2: The integrated stun server should only be used for debugging or development purposes. 

- Note 3: It will still work if both Addr0 and Addr1 are identical. Just not recommended.

- Note 4: Each address listens on both Port0 and Port1.


General Log Settings
--------------------

- **logConsoleColors** - Enable colors to be shown in console log. [true|false]
- **logConsoleDate** - Prepend date and time to console log. [true|false]


Service Log Settings
--------------------
Logs can be enabled can controlled independantly for the following services.

- **EasyRtc** - Logs server starts/stops and EasyRTC commands sent within websockets.
- **Express** - Logs http and https connection information.
- **SocketIo** - Logs generic websocket information. Not highly detailed.
- **Stun** - Logs STUN connection information if experimantal STUN server is enabled.


*Rather than repeat commands for every service, replace ??? with the service name.*

- **log???ConsoleEnabled** - Enable console logging [true|false]
- **log???ConsoleLevel** - Lowest priority level to start showing logs for. ['debug'|'info'|'warn'|'error']
- **log???FileEnabled** - Enable file logging. [true|false]
- **log???FileLevel** - Lowest priority level to start showing logs for. ['debug'|'info'|'warn'|'error']
- **log???FileName** - File location relative to the EasyRTC root.


Socket.io Settings
------------------
- **config.socketIoClientGzipEnabled** - Gzip socket.io client. Enabling could cause problems on some Windows installations. [true|false]
- **config.socketIoClientMinifyEnabled** - Pre-minify socket.io javascript. This is done just once and greatly saves on bandwidth. [true|false]
- **config.socketIoClientEtagEnabled** - Allow client caching of the socket.io javascript library [true|false]


Other Settings
--------------

- **defaultApplicationName** - The namespace for the default application. This can be read and reassigned via the EasyRTC API.
- **enableDemos** - Enable the provided EasyRTC demos. Viewable in /demos/. [true|false]
- **updateCheckEnable** - Checks in with easyrtc.com to see if a new version is available. [true|false]


docs/easyrtc_server_configuration.md
EasyRTC: Server Configuration
=============================

**Alpha Version**
*This page will be updated before we leave alpha, and certainly before we get to production. If not, feel free to attack us with wet noodles.*

For the code setting the default configuration options see:
 - [../lib/easyrtc_default_options.js](../lib/easyrtc_default_options.js)

Configuration levels - Server versus Application versus Room
----

In the near(ish) future, we will be exposing methods to manage configuration options at the application and room level. This will allow different rooms in the same application, or different applications in the same server to run with customized options.

Setting Server Options
----------------------

Options can be set prior, during, or after the easyrtc.listen() function has been run.

**Seperate**
 - easyrtc.setOption(optionName, optionValue)

**Within Listen()**
 - easyrtc.listen(httpApp, socketServer, options, listenCallback)

**Afterwards with resulting EasyRTC interface function**
 - ei.setOption(optionName, optionValue)

*note: Not all options will have an effect if changed after the server has started.*

Setting Server Options Example
------------------------------
The following server.js code snippet includes three ways of setting EasyRTC options.


	var http    = require("http");              // http server core module
	var express = require("express");           // web framework external module
	var io      = require("socket.io");         // web socket external module
	var easyrtc = require("easyrtc");           // EasyRTC external module
	
	// Setup and configure Express http server. Expect a subfolder called "static" to be the web root.
	var httpApp = express();
	httpApp.configure(function() {
	    httpApp.use(express.static(__dirname + "/static/"));
	});

	var webServer = http.createServer(httpApp).listen(8080);
	var socketServer = io.listen(webServer);

	
	// Configure EasyRTC to load demos from /easyrtcdemos/ 
	easyrtc.setOption("demosPublicFolder", "/easyrtcdemos");

	// Start EasyRTC server with options to change the log level and add dates to the log.
	var easyrtcServer = easyrtc.listen(
			httpApp,
			socketServer,
			{logLevel:"debug", logDateEnable:true},
			easyrtcListener
	);

	
	var easyrtcListener = function(err, ei) {
		// After the server has started, we can still change the default room name
		ei.setOption("roomDefaultName", "SectorZero");
	};

Available Server Options
------------------------

**WebRTC Options**
 - iceServers
   - ICE Servers object which identifies all STUN and TURN servers.
   - With new username field in TURN specification, this object may be updated soon.
   - Defaults to: [{url: "stun:stun.l.google.com:19302"},{url: "stun:stun.sipgate.net"},{url: "stun:217.10.68.152"},{url: "stun:stun.sipgate.net:10000"},{url: "stun:217.10.68.152:10000"}]


**Application Options**
 - appDefaultName
   - The default application a connection belongs to if it is not initially specified.
   - Defaults to: "default"
 - appAutoCreateEnable
   - Enables the creation of rooms from the API. Occurs when client joins a nonexistent room.
   - Defaults to: true             

**Room Options**
 - roomDefaultEnable    
   -  Enables connections joining a default room if it is not initially specified. If false, than a connection initially may be in no room.
   - Defaults to: true                        
 - roomDefaultName      
   - The default room a connection joins if it is not initially specified.
   - Defaults to: "default"                   
 - roomAutoCreateEnable 
   - Enables the creation of rooms from the API. Occurs when client joins a nonexistent room.
   - Defaults to: true


**Connection Options**
 - connectionDefaultField 
   - Default connection fields.
   - Defaults to: {browserFamily:  {share:true, regex:null, data:null},browserMajor:   {share:true, regex:null, data:null},osFamily:       {share:true, regex:null, data:null},deviceFamily:   {share:true, regex:null, data:null}}


**API Hosting Options**
 - apiEnable            
   - Enables hosting of the EasyRTC API files.
   - Defaults to: true                       
 - apiPublicFolder      
   - Api public folder without trailing slash. Note that the demos expect this to be '/easyrtc'
   - Defaults to: "/easyrtc"       
 - apiMinifyJsEnable    
   - Minify's the API Javascript for faster transfer.
   - NOT YET IMPLEMENTED
   - Defaults to: true                         
 - apiOldLocationEnable 
   - [Depreciated] Listens for requests to core API files in old locations (in addition to the new standard locations)
   - Defaults to: true         


**Demo Options**
 - demosEnable
   - Enables the various EasyRTC demos. For a production system, this should be disabled.
   - Defaults to: true
 - demosPublicFolder
   - Demos public folder without trailing slash.
   - Defaults to: "/demos"                      


**Log options** - Only apply if internal 'log' event is used
 - logLevel
   - The minimum log level to show. (debug|info|warning|error|none)
   - Defaults to: "info"
 - logDateEnable     
   -  Display timestamp in each entry
   - Defaults to: false
 - logErrorStackEnable 
   - print the stack trace in logged errors when available
   - Defaults to: true
 - logWarningStackEnable
   - print the stack trace in logged warnings when available
   - Defaults to: true
 - logColorEnable     
   - include console colors. Disable if forwarding logs to files or databases
   - Defaults to: true
 - logObjectDepth
   - When objects are included in the log, this is the max depth the log will display
   - Defaults to: 7


**Miscellaneous Server Options**
 - updateCheckEnable
   - Checks for updates
   - Defaults to: true


**Regular expressions for validating names and other input**
 - apiVersionRegExp
   - API Version
   - Defaults to: /^[a-z0-9_.-]{1,32}$/i
 - appNameRegExp
   - Application name 
   - Defaults to: /^[a-z0-9_.-]{1,32}$/i       
 - easyrtcidRegExp
   - EasyRTC socket id (easyrtcid)
   - Defaults to: /^[a-z0-9_.-]{1,32}$/i       
 - easyrtcsidRegExp
   - EasyRTC session id (easyrtcsid)
   - Defaults to: /^[a-z0-9_.-]{1,64}$/i       
 - groupNameRegExp
   - Group name
   - Defaults to: /^[a-z0-9_.-]{1,32}$/i       
 - fieldNameRegExp
   - Field names (for defining app and room custom fields)
   - Defaults to: /^[a-z0-9_. -]{1,32}$/i      
 - optionNameRegExp
   - Option names (for defining server options)
   - Defaults to: /^[a-z0-9_. -]{1,32}$/i      
 - presenceShowRegExp
   - Allowed presence "show" values (for setPresence command)
   - Defaults to: /^(away|chat|dnd|xa)$/       
 - presenceStatusRegExp
   - Allowed presence "status" value
   - Defaults to: /^(.){0,255}$/               
 - roomNameRegExp
   - Room name
   - Defaults to: /^[a-z0-9_.-]{1,32}$/i       
 - sessionKeyRegExp
   - Session key (easyrtcsid)
   - Defaults to: /^[a-z0-9_.-]{1,32}$/i       
 - usernameRegExp
   - Username
   - Defaults to: /^(.){1,64}$/i                



config.js
// easyRTC Configuration File
var config = {}; 

// User Configurable Options
// *************************
config.httpPort     = 8080;                 // The port which the http and socket server is on.
config.sslEnable    = false;                // SSL - Note: if https is enabled, http is disabled.
config.sslPort      = 8443;
config.sslForwardFromHttp = true;           // Forces non-secure connections to be forwarded to the secure site. (recommended true)
config.sslKeyFile   = '/ssl/mydomain.key';  // Local file location of private SSL key
config.sslCertFile  = '/ssl/mydomain.crt';  // Local file location of public SSL certificate


// External stun server (will be used if experimentalStunServerEnable is false)
// Several public stun servers are available to be used.
config.externalIceServers = [
    {url: "stun:stun.l.google.com:19302"},
    {url: "stun:stun.sipgate.net"},
    {url: "stun:217.10.68.152"},
    {url: "stun:stun.sipgate.net:10000"},
    {url: "stun:217.10.68.152:10000"}
];

// EXPERIMENTAL STUN SERVER
// It isn't good for production, but is nice for closed development.
// You "should" include two addresses which are reachable by all clients.
// Ports would need to be unblocked by firewalls.
config.experimentalStunServerEnable = false;  
config.experimentalStunServerAddr0  = "192.168.1.100";
config.experimentalStunServerAddr1  = "192.168.1.101";  // "should" be a seperate from Addr0
config.experimentalStunServerPort0  = "3478";           // Each IP listens on both Port0 and Port1
config.experimentalStunServerPort1  = "3479";


// Logging Settings
config.logConsoleColors         = true;             // Display colors in console log (affects all logs)
config.logConsoleDate           = false;            // Prepend date and time to console log (affects all logs)

// Logging: General easyRTC Server Logging
config.logEasyRtcConsoleEnabled = true;             // Enable easyRTC general console logging
config.logEasyRtcConsoleLevel   = 'info';           // ('debug'|'info'|'warn'|'error')

config.logEasyRtcFileEnabled    = false;            // Enable easyRTC general file logging
config.logEasyRtcFileLevel      = 'info';           // ('debug'|'info'|'warn'|'error')
config.logEasyRtcFileName       = ('./logs/easyrtc.log');// File location relative to easyRTC root

// Logging: Express Module
config.logExpressConsoleEnabled = false;            // Enable Express console logging
config.logExpressConsoleLevel   = 'info';           // ('debug'|'info'|'warn'|'error')

config.logExpressFileEnabled    = false;            // Enable Express file log
config.logExpressFileLevel      = 'info';           // ('debug'|'info'|'warn'|'error')
config.logExpressFileJson       = true;             // Enable JSON format in log file
config.logExpressFileName       = ('./logs/express.log');// File location relative to easyRTC root

// Logging: Socket.IO Module (not terribly useful)
config.logSocketIoConsoleEnabled= true;             // Enable Socket.io console logging
config.logSocketIoConsoleLevel  = 'warn';           // ('debug'|'info'|'warn'|'error')

config.logSocketIoFileEnabled   = false;            // Enable Socket.io file logging
config.logSocketIoFileLevel     = 'warn';           // ('debug'|'info'|'warn'|'error')
config.logSocketIoFileName      = ('./logs/socketio.log'); // File location relative to easyRTC root

// Logging: Experimental Stun Module
config.logStunConsoleEnabled    = true;             // Enable STUN server console logging
config.logStunConsoleLevel      = 'info';           // ('debug'|'info'|'warn'|'error')

config.logStunFileEnabled       = false;            // Enable STUN server file logging
config.logStunFileLevel         = 'warn';           // ('debug'|'info'|'warn'|'error')
config.logStunFileName          = ('./logs/stun.log');// File location relative to easyRTC root


// The namespace for the default application.
config.defaultApplicationName = "default";

// Enable easyRTC demos. Viewable in /demos/
config.enableDemos = true;

// Socket.IO Settings
config.socketIoClientGzipEnabled    = false;        // Gzip socket.io client. Enabling could cause problems on some Windows installations.
config.socketIoClientMinifyEnabled  = true;         // Pre-minify socket.io javascript. This is done just once and greatly saves on bandwidth.
config.socketIoClientEtagEnabled    = true;         // Allow client caching of the socket.io javascript library

// Check for updates
config.updateCheckEnable = true;

// **********************************
// End of user configurable settings.
// **********************************

// The following settings work alongside the easyRTC client API.
// DO NOT ALTER!
// *************************************************************

// Load version number from package.json file 
var easyrtcPackage = require('./package');
config.easyrtcVersion = easyrtcPackage.version;
delete require.cache[easyrtcPackage];

config.cmdPacketType = "easyrtcCmd";
config.cmdMsgType = {
    error: "error",
    list: "list",
    token: "token"
};

// Programmatically Setting Some Settings
// **************************************
if (config.experimentalStunServerEnable)
    config.iceServers = [
        {"url": "stun:" + config.experimentalStunServerAddr0 + ":" + config.experimentalStunServerPort0},
        {"url": "stun:" + config.experimentalStunServerAddr0 + ":" + config.experimentalStunServerPort1},
        {"url": "stun:" + config.experimentalStunServerAddr1 + ":" + config.experimentalStunServerPort0},
        {"url": "stun:" + config.experimentalStunServerAddr1 + ":" + config.experimentalStunServerPort1}
    ];
else
    config.iceServers = config.externalIceServers;

// Allows the config object to be seen by a caller
module.exports = config;
